### ISSUE TITLE
Field defaults evaluated when encodeDefaults = false

### ISSUE THREAD

**Issue by ejektaflex** [Kotlin/kotlinx.serialization#1288]
**Describe the bug**

When `encodeDefaults = true`, the value of a field is evaluated BEFORE checking whether or not to actually encode the default value. This should happen afterwards. Currently, this causes defaults to be evaluated even though it's not required, and if the field has a default value that changes (such as in the example code below), it will cause that value to be re-evaluated.

I thought that `encodeDefaults` existed to help avoid the complications that arose from evaluating default values multiple times, but it seems that this bug does not allow that to happen.

**To Reproduce**

Here is a simplified version of the code that I am running:

```kotlin
val PawnRegistry = GameRegistry("pawn", Pawn::class)

@Serializable
class Pawn : Idable<Pawn> {
    @Transient 
    override val reg = PawnRegistry
    override val id = reg.addEntry(this)
}
```
As you can see, we set a default value for `id`. However, EVERY time this Pawn gets serialized, it gets added to the game registry again, *whether or not `encodeDefaults` is true*. The reason why comes from the bytecode that is emitted from Kotlinx Serialization during `public final static write$Self`:
```assembly
   L4
    LINENUMBER 25 L4
    ALOAD 0
    INVOKEVIRTUAL io/ejekta/models/core/Pawn.getReg ()Lio/ejekta/api/struct/GameRegistry;
    ALOAD 0
    CHECKCAST io/ejekta/api/struct/Idable
    ACONST_NULL
    ICONST_2
    ACONST_NULL
    INVOKESTATIC io/ejekta/api/struct/GameRegistry.addEntry$default (Lio/ejekta/api/struct/GameRegistry;Lio/ejekta/api/struct/Idable;Ljava/lang/String;ILjava/lang/Object;)Ljava/lang/String;
    INVOKESTATIC kotlin/jvm/internal/Intrinsics.areEqual (Ljava/lang/Object;Ljava/lang/Object;)Z
    ICONST_1
    IXOR
    IFNE L5
    ALOAD 1
    ALOAD 2
    ICONST_1
    INVOKEINTERFACE kotlinx/serialization/encoding/CompositeEncoder.shouldEncodeElementDefault (Lkotlinx/serialization/descriptors/SerialDescriptor;I)Z (itf)
    IFEQ L6
   L5
```

**Expected behavior**

I wouldn't expect the Kotlinx Serialization to be evaluating defaults if it doesn't even encode them. Why check whether or not the field is equal to it's default value, if we don't even need to encode it? This makes serializing data with non-constant default values *impossible*, since each time you check the value, it changes. It needs to first check whether or not we need to encode the element, and then and *only then* should it check the value.

To fix, I'd recommend checking `CompositeEncoder.shouldEncodeElementDefault` *before* comparing defaults with `Intrinsics.areEqual`. We shouldn't be checking if the defaults are equal if we don't even need to encode the field.

**Environment**
 - Kotlin version: 1.4.20
 - Library version: 1.4.10
 - Kotlin platforms: JVM
 - Gradle version: [e.g. 4.10]
 - Windows 10

**Comment by shanshin** [comment]
>I thought that encodeDefaults partially existed to help avoid the complications that arose from evaluating default values multiple times, but it seems that this bug does not allow that to happen.

The main reasons were other https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/basic-serialization.md#defaults-are-not-encoded
In any case properties with calculated default values are not fully supported now (see #133) so this problem will be researched.

**Comment by ejektaflex** [comment]
Yes - I do understand the reasoning better now 👍 . This is more to point out that regardless of reasoning, we needn't always re-evaluate property defaults, as this can lead to very inconsistent hidden behaviour. Thank you for researching it!

**Comment by pdvrieze** [comment]
Independent of the wisdom of having auto-self-registering values (Pawns), the code change you highlight is worthwhile, so you should create a pull request for it. On the other hand, there is no guarantee that the implementation of `shouldEncodeElementDefault` is actually cheap (even though the current/common implementation is a simple field lookup for most formats).

Your suggested code I think has some logic errors (but it depends on the surrounding code): `if (shouldEncodeElementDefault(field) || field == field.default_value) { /*encode value*/ }`

**Comment by bvisness** [comment]
I don't think performance is necessarily such a big concern, especially since I would never really expect `shouldEncodeElementDefault` to be expensive (at least if it's just looking in some configuration data). I think the bigger issue is the fact that checking defaults in this way can have side effects, and currently there's no way to suppress them.

I question whether it ever makes sense to suppress encoding of default values when the default is not a compile-time constant. If the default value could be different every time it's called, I don't think the optimization makes sense at all.

**Comment by shanshin** [comment]
Fixed in Kotlin 1.5.0-RC
